### PR TITLE
[c++ grpc] Fix two server-side crashes

### PR DESCRIPTION
**Pull request by chwarr**
Two server-side crashes were revealed by the compatibility testing:

* sometimes, when the server was being shutdown, in-flight RPC calls couldn't send back responses
* the state of a `unary_call` was sometimes being deleted before the `unary_call` instance was itself destroyed

There are two separate commits here, but both are needed to fix the server-side crashes, hence one PR.

Please `rebase -i` and fixup when merging to maintain the commits.

### CODE REVIEW

File: cpp/inc/bond/ext/grpc/detail/service_call_data.h [hunk 1]
```diff
@@ -81,18 +81,24 @@ struct service_unary_call_data : io_manager_tag
...
-            // back to the completion queue as the result of sending a
-            // response. The UnaryCall wrapper that we create inside the
-            // thread guarantees that some response will always be sent.
-            unary_call_impl<TRequest, TResponse>* receivedCall = _receivedCall.release();
-
-            _threadPool->schedule([this, receivedCall]()
             {
-                _cb(unary_call<TRequest, TResponse> { receivedCall });
-            });
+                // capture the data associated with this one incomming
```
Comment (chwarr) [inline cpp/inc/bond/ext/grpc/detail/service_call_data.h]: Weird diff here. Let me add a comment about the scope block.

File: cpp/inc/bond/ext/grpc/unary_call.h [hunk 2]
```diff
@@ -20,31 +20,70 @@
...
+    /// The user-facing unary_call object also has a shared_ptr to an
+    /// instance of unary_call_impl to keep the state alive while the call
+    /// object is itself alive.
     template <typename TRequest, typename TResponse>
-    struct unary_call_impl final : detail::io_manager_tag
+        struct unary_call_impl final : std::enable_shared_from_this<unary_call_impl<TRequest, TResponse>>
     {
+        /// @brief Handles keeping a unary_call_impl instance alive while a
+        /// response it being sent.
+        struct response_sent_tag final : detail::io_manager_tag
```
Comment (chwarr) [inline cpp/inc/bond/ext/grpc/unary_call.h]: I added this class to attempt to separate the `unary_call_impl` data from the response sending behavior. Let me know whether or not you think this makes things clearer or muddier.

File: cpp/inc/bond/ext/grpc/detail/service_call_data.h (hunk 1 above)
Comment (chwarr) [inline cpp/inc/bond/ext/grpc/detail/service_call_data.h]: Fixed.

File: cpp/inc/bond/ext/grpc/detail/service_call_data.h [hunk 3]
```diff
@@ -69,7 +69,7 @@ struct service_unary_call_data : io_manager_tag
           _cq(cq),
           _threadPool(threadPool),
           _cb(cb),
-          _receivedCall(new unary_call_impl<TRequest, TResponse>)
+        _receivedCall(std::make_shared<unary_call_impl<TRequest, TResponse>>())
```
Comment (lalo) [inline cpp/inc/bond/ext/grpc/detail/service_call_data.h]: nit: missing one space

File: cpp/inc/bond/ext/grpc/unary_call.h [hunk 4]
```diff
@@ -20,31 +20,70 @@
...
+    /// itself that it enques in the completion queue. When that sending of
+    /// the response is done and it is dequeued from the completion queue,
+    /// response_sent_tag clears its shared_ptr to the unary_call_impl.
+    ///
+    /// The user-facing unary_call object also has a shared_ptr to an
+    /// instance of unary_call_impl to keep the state alive while the call
+    /// object is itself alive.
     template <typename TRequest, typename TResponse>
-    struct unary_call_impl final : detail::io_manager_tag
+        struct unary_call_impl final : std::enable_shared_from_this<unary_call_impl<TRequest, TResponse>>
```
Comment (lalo) [inline cpp/inc/bond/ext/grpc/unary_call.h]: nit: spaces? or is this common c++

Comment (chwarr) [inline cpp/inc/bond/ext/grpc/unary_call.h]: No. It's Visual Studio and emacs fighting with each other...

File: cpp/inc/bond/ext/grpc/detail/service_call_data.h [hunk 5]
```diff
@@ -69,7 +69,7 @@ struct service_unary_call_data : io_manager_tag
           _cq(cq),
           _threadPool(threadPool),
           _cb(cb),
-        _receivedCall(std::make_shared<unary_call_impl<TRequest, TResponse>>())
+         _receivedCall(std::make_shared<unary_call_impl<TRequest, TResponse>>())
```
Comment (lalo) [inline cpp/inc/bond/ext/grpc/detail/service_call_data.h]: nit: one more space

Comment (chwarr) [inline cpp/inc/bond/ext/grpc/detail/service_call_data.h]: Ack. Pending locally.

**Comment by chwarr** [comment]
Merged in commit f6c065cf.

**Pull request closed without merging**